twilight-gateway/src/ratelimiter.rs

```diff
@@ -1,127 +1,240 @@
 //! Ratelimiter on the user's ability to [send messages].
 //!
+//! See <https://discord.com/developers/docs/topics/gateway#rate-limiting>
+//!
 //! [send messages]: crate::Shard::send
 
-use leaky_bucket_lite::LeakyBucket;
-use std::time::{Duration, Instant};
+use std::{
+    future::{poll_fn, Future},
+    pin::Pin,
+    task::{ready, Context, Poll},
+};
+use tokio::time::{sleep, Duration, Instant, Sleep};
+
+/// Number of commands allowed in a given [`RESET_DURATION`].
+const COMMANDS_PER_RESET: u8 = 120;
 
-/// Interval of how often the ratelimit bucket resets, in milliseconds.
-const RESET_DURATION_MILLISECONDS: u64 = 60_000;
+/// Duration until the ratelimit bucket resets.
+const RESET_DURATION: Duration = Duration::from_secs(60);
 
 /// Ratelimiter for sending commands over the gateway to Discord.
 #[derive(Debug)]
 pub struct CommandRatelimiter {
-    /// Bucket used for limiting actions.
-    bucket: LeakyBucket,
+    /// Future that completes the next time the `CommandRatelimiter` allows a
+    /// permit.
+    delay: Pin<Box<Sleep>>,
+    /// Queue of instants started when a command was sent.
+    instants: Vec<Instant>,
 }
 
 impl CommandRatelimiter {
-    /// Create a new ratelimiter.
-    pub(crate) fn new(heartbeat_interval: u64) -> Self {
-        /// Interval of how often to refill the bucket.
-        const REFILL_INTERVAL: Duration = Duration::from_millis(RESET_DURATION_MILLISECONDS);
-
-        // Number of commands allotted to the user per reset period.
-        let commands_allotted = u32::from(available_commands_per_interval(heartbeat_interval));
-
-        let bucket = LeakyBucket::builder()
-            .max(commands_allotted)
-            .tokens(commands_allotted)
-            .refill_interval(REFILL_INTERVAL)
-            .refill_amount(commands_allotted)
-            .build();
-
-        Self { bucket }
+    /// Create a new ratelimiter with some capacity reserved for heartbeating.
+    pub(crate) fn new(heartbeat_interval: Duration) -> Self {
+        let allotted = nonreserved_commands_per_reset(heartbeat_interval);
+
+        Self {
+            delay: Box::pin(sleep(Duration::ZERO)),
+            instants: Vec::with_capacity(allotted.into()),
+        }
     }
 
     /// Current number of commands that are still available within the interval.
-    pub fn available(&self) -> u32 {
-        self.bucket.tokens()
+    pub fn available(&self) -> u8 {
+        // filter out elapsed instants
+        let used_permits: u8 = self
+            .instants
+            .iter()
+            .filter(|instant| instant.elapsed() < RESET_DURATION)
+            .count()
+            .try_into()
+            .expect("length is at most 118");
```

Comment:
sorry for the runaround, but i like this less. i don't want to use `unwrap` or `expect` if it can be avoided. i suppose ignoring the clippy lint is indeed better